crates/storage/provider/src/providers/snapshot/manager.rs

```diff
@@ -449,6 +449,91 @@ impl SnapshotProvider {
     pub fn directory(&self) -> &Path {
         &self.path
     }
+
+    /// Retrieves data from the database or snapshot, wherever it's available.
+    ///
+    /// # Arguments
+    /// * `segment` - The segment of the snapshot to check against.
+    /// * `index_key` - Requested index key, usually a block or transaction number.
+    /// * `fetch_from_snapshot` - A closure that defines how to fetch the data from the snapshot
+    ///   provider.
+    /// * `fetch_from_database` - A closure that defines how to fetch the data from the database
+    ///   when the snapshot doesn't contain the required data or is not available.
+    pub fn get_with_snapshot_or_database<T, FS, FD>(
```

Comment:
this a duplicate from another function which was part of `DatabaseProvider`, will dedup on a follow-up